proto/src/proto.rs

```diff
@@ -372,7 +372,32 @@ pub enum LdapDerefAliases {
 pub struct LdapSubstringFilter {
     pub initial: Option<String>,
     pub any: Vec<String>,
-    pub final_: Option<String>,
+    pub final_: Option<String>, //escape final feyword
+}
+
+impl From<String> for LdapSubstringFilter {
+    fn from(value: String) -> Self {
+        let mut filter = LdapSubstringFilter {
+            initial: None,
+            any: Vec::new(),
+            final_: None,
+        };
+
+        let parts: Vec<&str> = value.split('*').collect();
+
+        for (i, part) in parts.iter().enumerate() {
+            match i {
+                0 if !part.is_empty() => filter.initial = Some(part.to_string()),
+                _ if i == parts.len() - 1 && !part.is_empty() => {
+                    filter.final_ = Some(part.to_string())
+                }
+                _ if !part.is_empty() => filter.any.push(part.to_string()),
+                _ => (),
+            }
+        }
```

Comment:
Issue (non-blocking): it appears the first element is special cased by using `enumerate` and `match`.

Suggestion: not perfect by any means, but you could do something like this instead:
```rust
let n_separators = value.split('*').count();
let mut segments = value.split('*');

filter.initial = segments.next().filter(|s| !s.is_empty()).map(|s| s.to_owned());

if 1 <= n_separators {
    segments.by_ref().take(n_separators - 1).filter(|s| !s.is_empty()).for_each(|s| filter.any.push(s.to_owned()));
}

filter.final_ = segments.next().filter(|s| !s.is_empty()).map(|s| s.to_owned());
```

Reply:
I get something that does not allocate memory, which I think is easier to read. let me know what you think